src/lib/optimizer/strategy/predicate_split_up_rule.cpp

```diff
@@ -2,44 +2,91 @@
 
 #include "expression/expression_utils.hpp"
 #include "expression/logical_expression.hpp"
-#include "expression/lqp_subquery_expression.hpp"
 #include "logical_query_plan/lqp_utils.hpp"
+#include "logical_query_plan/union_node.hpp"
 
 namespace opossum {
 
+PredicateSplitUpRule::PredicateSplitUpRule(const bool split_disjunctions) : _split_disjunctions(split_disjunctions) {}
+
 void PredicateSplitUpRule::apply_to(const std::shared_ptr<AbstractLQPNode>& root) const {
   Assert(root->type == LQPNodeType::Root, "PredicateSplitUpRule needs root to hold onto");
 
-  /**
-   * Step 1:
-   *    - Collect PredicateNodes that can be split up into multiple ones into `predicate_nodes_to_flat_conjunctions`
-   */
-  auto predicate_nodes_to_flat_conjunctions =
-      std::vector<std::pair<std::shared_ptr<PredicateNode>, std::vector<std::shared_ptr<AbstractExpression>>>>{};
-
+  auto predicate_nodes = std::vector<std::shared_ptr<PredicateNode>>{};
   visit_lqp(root, [&](const auto& sub_node) {
     if (const auto predicate_node = std::dynamic_pointer_cast<PredicateNode>(sub_node)) {
-      const auto flat_conjunction = flatten_logical_expressions(predicate_node->predicate(), LogicalOperator::And);
-
-      if (flat_conjunction.size() > 1) {
-        predicate_nodes_to_flat_conjunctions.emplace_back(predicate_node, flat_conjunction);
-      }
+      predicate_nodes.emplace_back(predicate_node);
     }
-
     return LQPVisitation::VisitInputs;
   });
 
+  // _split_conjunction() and _split_disjunction() split up logical expressions by calling each other recursively.
+  for (const auto& predicate_node : predicate_nodes) {
+    const auto flat_conjunction = flatten_logical_expressions(predicate_node->predicate(), LogicalOperator::And);
+    if (flat_conjunction.size() > 1) {
```

Comment:
You are right, this code is shorter and works as well. We now might have a superfluous call of `flatten_logical_expressions()` in `_split_disjunction()` because `_split_conjunction()` already unnested all expressions, but this does not increase the runtime compared to the previous solution.